crates/apollo-compiler/src/coordinate.rs

```diff
@@ -0,0 +1,286 @@
+use crate::ast::InvalidNameError;
+use crate::schema::Name;
+use std::fmt;
+use std::str::FromStr;
+
+/// Create a static coordinate at compile time.
+///
+/// ```rust
+/// use apollo_compiler::coord;
+///
+/// assert_eq!(coord!(@directive).to_string(), "@directive");
+/// assert_eq!(coord!(@directive(arg:)).to_string(), "@directive(arg:)");
+/// assert_eq!(coord!(Type).to_string(), "Type");
+/// assert_eq!(coord!(Type.field).to_string(), "Type.field");
+/// assert_eq!(coord!(Type.field(arg:)).to_string(), "Type.field(arg:)");
+/// ```
+#[macro_export]
+macro_rules! coord {
+    ( @ $name:ident ) => {
+        $crate::coordinate::DirectiveCoordinate($crate::name!($name))
+    };
+    ( @ $name:ident ( $arg:ident : ) ) => {
+        $crate::coordinate::DirectiveArgumentCoordinate($crate::name!($name), $crate::name!($arg))
+    };
+    ( $name:ident ) => {
+        $crate::coordinate::TypeCoordinate($crate::name!($name))
+    };
+    ( $name:ident . $field:ident ) => {
+        $crate::coordinate::FieldCoordinate($crate::name!($name), $crate::name!($field))
+    };
+    ( $name:ident . $field:ident ( $arg:ident : ) ) => {
+        $crate::coordinate::FieldArgumentCoordinate(
+            $crate::name!($name),
+            $crate::name!($field),
+            $crate::name!($arg),
+        )
+    };
+}
+
+/// A schema coordinate targeting a type definition: `Type`.
+///
+/// # Example
+/// ```
+/// use apollo_compiler::name;
+/// use apollo_compiler::coordinate::TypeCoordinate;
+///
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// assert_eq!(TypeCoordinate(name!("Type")).to_string(), "Type");
+/// # Ok(()) }
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TypeCoordinate(pub Name);
```

Comment:
I’d prefer named fields for all of these structs, especially for those with 2+ fields of the same type (but let’s keep them consistent with each other)